### PR TITLE
Add prepare_dataset command

### PR DESCRIPTION
**Pull request by tscholak**
# ✨ Description

Extracted and refined the dataset preparation script from #17.
Made it a command like `train` or `convert`.
Example call and config:

```bash
fast-llm prepare_dataset gpt_memmap --config foo.yaml
```

or

```
torchrun --standalone --nnodes 1 --nproc_per_node=1 --no_python \
    fast-llm prepare_dataset gpt_memmap --config foo.yaml
```

where `foo.yaml` contains:

```yaml
output_path: /tmp/foo

loading_workers: 4
tokenize_workers: 4
saving_workers: 4

dataset:
  name_or_path: stas/openwebtext-10k

tokenizer:
  path: /tmp/SmolLM-135M/tokenizer.json
```

Run `git clone https://huggingface.co/HuggingFaceTB/SmolLM-135M` in `tmp` to get that tokenizer file.

This will produce:

```
/tmp/foo
├── downloaded_dataset
│   ├── cache-1e5559f36da9962e_00002_of_00004.arrow
│   ├── cache-1e5559f36da9962e_00003_of_00004.arrow
│   ├── cache-1e5559f36da9962e_00001_of_00004.arrow
│   ├── cache-1e5559f36da9962e_00000_of_00004.arrow
│   ├── data-00001-of-00004.arrow
│   ├── dataset_info.json
│   ├── data-00000-of-00004.arrow
│   ├── data-00002-of-00004.arrow
│   ├── data-00003-of-00004.arrow
│   ├── ok
│   └── state.json
├── shard_0_0.idx
├── shard_0_0.bin
└── fast_llm_dataset.json
```

with `fast_llm_dataset.json` reading:

```json
{
    "datasets": [
        {
            "prefix": "shard_0_0",
            "num_documents": 10000,
            "num_tokens": 11569536,
            "weight": 1.0
        }
    ]
}
```

The `downloaded_dataset` can be deleted afterwards. It is not used by Fast-LLM.

## 🔍 Type of change

Select all that apply:

- [x] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [x] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

1. Fixed memory-mapped indexed dataset and added round-trip tests
2. Added `prepare_dataset` command
3. Simplified `Dockerfile`

# ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General:
- [x] 📜 I have read and followed the [contributing guidelines](CONTRIBUTING.md).
- [x] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [x] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration:
- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing:
- [x] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact:
- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

# 📊 Performance Impact Details

N/A

---

# 📝 Additional Notes

N/A

### CODE REVIEW

**Comment by tscholak** [comment]
Note to self: don't forget to enforce the use of hf_transfer for downloading

**Comment by tscholak** [comment]
@jlamypoirier 

I introduced some classes so that we can have different data preparations for different models. What we have there now is only appropriate for training gpts but won't work for vlms. I'm planning with @akshaykalkunte to add preparation code for StarDoc.

About the changes to the Dockerfile: I realized that we can install Fast-LLM in editable mode globally if we make /app writable for everyone. I therefore removed the fast-llm used. This also resolves a problem where fast-llm isn't usable if the user's id in the job doesn't match the id of the fast-llm user.

File: fast_llm/tools/prepare_dataset.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+from fast_llm.data.config import TokenizerConfig
+from fast_llm.data.gpt.memmap import GPTMemmapDataset
+from fast_llm.data.tokenizer import Tokenizer
+from fast_llm.engine.config_utils.data_type import DataType
+from fast_llm.engine.config_utils.runnable import RunnableConfig
+from fast_llm.utils import Assert, Registry
+
+
+@config_class
+class DistributedConfig(Config):
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:23]: This is basically a duplicate of the other distributed config, are you really sure we can't merge them? What would be the minimal requirement requirement for combining them?

File: fast_llm/tools/prepare_dataset.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        if self.world_size is None:
+            self.world_size = self.default_world_size
+        if self.rank is None:
+            self.rank = self.default_rank
+        super()._validate()
+        Assert.in_range(self.rank, 0, self.world_size)
+
+
+@config_class()
+class DatasetPreparatorConfig(RunnableConfig):
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:55]: Could we move dataset preparators to a separate module? Maybe `fast_llm/data/prepare/`? (and put config and gpt in separate files)

File: fast_llm/tools/prepare_dataset.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    _abstract = True
+    model_name: typing.ClassVar[str]
+
+    output_path: pathlib.Path = Field(
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:59]: Aren't these not fully generic?

File: fast_llm/tools/prepare_dataset.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    def get_dataset_preparator_class(cls) -> typing.Type["DatasetPreparator"]:
+        raise NotImplementedError
+
+    def _get_runnable(self, parsed: argparse.Namespace) -> typing.Callable[[], None]:
+        dataset_preparator = self.get_dataset_preparator_class()(config=self)
+        return dataset_preparator.run
+
+
+class DatasetPreparator(abc.ABC):
+    _abstract = True
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:79]: Why this?

File: fast_llm/tools/prepare_dataset.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    _config: DatasetPreparatorConfig
+    config_class: typing.ClassVar[type[DatasetPreparatorConfig]] = DatasetPreparatorConfig
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:81]: Is this used?

File: fast_llm/tools/prepare_dataset.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        default=False,
+        desc="Disable disk space check. Useful for environments where disk space is not accurately reported.",
+        hint=FieldHint.optional,
+    )
+
+
+@config_class()
+class GPTDatasetPreparatorConfig(DatasetPreparatorConfig):
+    _abstract = False
+    model_name: typing.ClassVar[str] = "gpt"
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:133]: I don't think we have a 1-1 match between datasets and models. The new dataset stuff leaves room for other GPT dataset formats, and there can be plenty of other scenarios for other models.

So I think we should decouple preparators from models. This one would have `preparator_name = "gpt_memmap"` and class `GPTMemmapDatasetPreparator[Config]`

File: fast_llm/tools/prepare_dataset.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+
+    tokens_per_shard: int = Field(
+        default=1_000_000_000,
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:136]: 10**9?

File: fast_llm/tools/prepare_dataset.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    )
+    _tokenizer: Tokenizer = Field(
+        init=False,
+        desc="The tokenizer instance.",
+        hint=FieldHint.derived,
+    )
+
+    def _validate(self):
+        Assert.not_none(self.tokenizer.path)
+        self._tokenizer = Tokenizer(config=self.tokenizer)
```
Comment (jlamypoirier) [inline fast_llm/tools/prepare_dataset.py:182]: This is doing some work and reading files, so it shouldn't go in `_validate` (not runnable on remote and requires dependencies). The tokenizer can be built in the preparator.

File: tests/test_memmap_dataset.py
Comment (jlamypoirier) [inline tests/test_memmap_dataset.py:1]: We should add at least 2 more tests, otherwise future changes will definitely break it:
* An `test_validate_without_import` for prepare dataset in `test_config`
* A test for `prepare_dataset`

**Review by jlamypoirier — CHANGES_REQUESTED** [review]

File: fast_llm/tools/prepare_dataset.py (hunk 8 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py:182]: Hm, not sure how I can both, derive the data type based on the tokenizer length and not import the tokenizer 😉
I can make it such that if the tokenizer dependency cannot be fulfilled the validation fails unless the data type is user specified in the config.

File: fast_llm/tools/prepare_dataset.py (hunk 7 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py:136]: Isn't that a float?

File: fast_llm/tools/prepare_dataset.py (hunk 5 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py:81]: I don't know. I find the machinery you built around these config classes very hard to follow.
I'll try to remove these things and hope I won't break things.

File: fast_llm/tools/prepare_dataset.py (hunk 2 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py:55]: Great, yes! I was hoping you'd tell me where to put this

File: Dockerfile [hunk 9]
```diff
@@ -7,28 +7,25 @@ RUN apt-get update \
...
+COPY ./tests tests
+COPY ./tools tools
+
+# Copy the main source code
+COPY --exclude=./fast_llm/csrc/ ./fast_llm/ fast_llm/
 
-# Copy the main source code for Fast-LLM
-COPY --exclude=./fast_llm/csrc/ --chown=fast_llm ./fast_llm/ fast_llm/
+# Ensure the source code files are writable
+RUN chmod -R a+w /app
```
[the remark this answers is not in the document]
Comment (tscholak) [reply Dockerfile:31]: Please see https://github.com/ServiceNow/Fast-LLM/pull/38#issuecomment-2468184561
There are bug issues right now which makes the current images built by the CI action unusable on Toolkit.

File: fast_llm/tools/prepare_dataset.py (hunk 1 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py:23]: See https://github.com/ServiceNow/Fast-LLM/pull/38#issuecomment-2468184561
I need help in unifying these and remove any hard dependency on GPUs and CUDA from what we have in Fast-LLM

File: tests/test_memmap_dataset.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+from hypothesis import given, strategies as st
+from hypothesis.extra import numpy as npst
+import numpy as np
+from tempfile import TemporaryDirectory
+from pathlib import Path
+from fast_llm.data.gpt.memmap import GPTMemmapDataset
+import pytest
+
+def dtype_arrays(dtype: np.dtype, min_size: int=1, max_size: int=100) -> st.SearchStrategy:
```
[the remark this answers is not in the document]
Comment (tscholak) [reply tests/test_memmap_dataset.py:9]: The benefit is that hypothesis will try to shrink the inputs to the minimal reproducible value in case of a problem

File: fast_llm/tools/prepare_dataset.py (hunk 1 above)
Comment (jlamypoirier) [reply fast_llm/tools/prepare_dataset.py:23]: Which hard dependencies? From what I see there is already a CPU-only option, but it comes with an `assert world_size == 1`. Apart from that, I think only the default nccl backend is a problem. Is there anything else?

For now, a compromise would be to use `DistributedConfig` but maybe not `Distributed`. The default backend could be changed with a FieldUpdate in a derived class, and ignoring a bunch of config parameters. What do you think?

File: fast_llm/tools/prepare_dataset.py (hunk 5 above)
Comment (jlamypoirier) [reply fast_llm/tools/prepare_dataset.py:81]: It's mostly ad-hoc, I haven't found a good way to standardize yet...

File: fast_llm/tools/prepare_dataset.py (hunk 7 above)
Comment (jlamypoirier) [reply fast_llm/tools/prepare_dataset.py:136]: 10e9 is a float, but not 10**9

File: fast_llm/tools/prepare_dataset.py (hunk 8 above)
Comment (jlamypoirier) [reply fast_llm/tools/prepare_dataset.py:182]: Can we do validation without knowing them? I think we could have `data_type: DataType | None = None` and derive the actual value in `run`.

File: Dockerfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (tscholak) [reply Dockerfile]: simplified the Dockerfile as discussed.

Build time is unaffected compared to `main`.

```
root@1caffc2c55ae:/app# ls -lah
total 64K
drwxr-xr-x+ 1 root root 4.0K Nov 12 12:57 .
drwxr-xr-x  1 root root 4.0K Nov 12 13:18 ..
drwxr-xr-x+ 2 root root 4.0K Oct  8 20:17 Megatron-LM
drwxrwxrwx+ 4 root root 4.0K Nov 12 12:57 build
drwxr-xr-x+ 2 root root 4.0K Nov 11 20:49 examples
drwxr-xr-x+ 1 root root 4.0K Nov 11 23:45 fast_llm
drwxrwxrwx+ 2 root root 4.0K Nov 12 12:57 fast_llm.egg-info
-rwxrwxrwx  1 root root   58 Oct  8 20:17 pyproject.toml
-rwxrwxrwx  1 root root 1.6K Nov 11 20:05 setup.cfg
-rwxrwxrwx  1 root root  877 Nov  9 16:21 setup.py
drwxr-xr-x+ 2 root root 4.0K Nov 11 23:45 tests
drwxr-xr-x+ 2 root root 4.0K Nov 11 20:49 tools
drwxrwxrwx+ 1 root root 4.0K Nov 12 12:56 venv
```

top-level `COPY`'ed folders are 755, but their contents are all 777:

```
root@1caffc2c55ae:/app# ls -lah fast_llm
total 140K
drwxr-xr-x+  1 root root 4.0K Nov 11 23:45 .
drwxr-xr-x+  1 root root 4.0K Nov 12 12:57 ..
-rwxrwxrwx   1 root root   22 Nov  3 18:06 __init__.py
-rwxrwxrwx   1 root root  34K Nov  3 18:06 config.py
drwxrwxrwx+  2 root root 4.0K Nov 11 23:12 core
drwxrwxrwx+  1 root root 4.0K Nov 11 23:12 csrc
drwxrwxrwx+  3 root root 4.0K Nov 11 23:45 data
drwxrwxrwx+ 11 root root 4.0K Nov 11 23:12 engine
drwxrwxrwx+  3 root root 4.0K Nov 11 23:12 functional
drwxrwxrwx+  5 root root 4.0K Nov 11 23:12 layers
-rwxrwxrwx   1 root root  11K Oct 16 14:52 logging.py
drwxrwxrwx+  4 root root 4.0K Nov 11 23:12 models
-rwxrwxrwx   1 root root  13K Oct 16 14:52 profile.py
-rwxrwxrwx   1 root root  12K Nov  9 16:18 tensor.py
drwxrwxrwx+  2 root root 4.0K Nov 11 23:45 tools
-rwxrwxrwx   1 root root 5.8K Nov 11 23:32 utils.py
```

File: fast_llm/tools/prepare_dataset.py (hunk 3 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py]: not sure what you mean?

File: fast_llm/tools/prepare_dataset.py (hunk 5 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py]: `_abstract` is not used, I removed it.
`config_class` is used in a validation check.

File: fast_llm/tools/prepare_dataset.py (hunk 4 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py]: removed

File: fast_llm/tools/prepare_dataset.py (hunk 1 above)
Comment (tscholak) [reply fast_llm/tools/prepare_dataset.py]: Let's leave this for another PR please.

**Review by jlamypoirier — APPROVED** [review]
LGTM, assuming my proposed modifications are ok
